matchzoo/generators/pair_generator.py

```diff
@@ -0,0 +1,120 @@
+        >>> task = tasks.Ranking()
+        >>> from matchzoo.generators import PairGenerator
+        >>> generator = PairGenerator(input, 1, task, 1, True)
+        >>> assert generator
+        >>> x, y = generator[0]
+
+    """
+
+    def __init__(
+        self,
+        inputs: datapack.DataPack,
+        num_neg: int=1,
+        task: engine.BaseTask=tasks.Ranking(),
+        batch_size: int=32,
+        shuffle: bool=True
+    ):
+        """Construct the pair generator.
+
+        :param inputs: the output generated by :class:`DataPack`.
+        :param num_neg: the number of negative samples associated with each
+            positive sample.
+        :param task: the task is a instance of :class:`engine.BaseTask`.
+        :param batch_size: number of instances in a batch.
+        :param shuffle: whether to shuffle the instances while generating a
+            batch.
+        """
+        self._task = task
+        self._num_neg = num_neg
+        self._column_names = inputs.dataframe.columns
+        self.pairs = self.transform_data(inputs)
+        super().__init__(batch_size, len(self.pairs), shuffle)
+
+    def transform_data(self, inputs: datapack.DataPack) -> dict:
+        """Obtain the transformed data from :class:`DataPack`.
+
+        Note here, label is required to make pairs.
+
+        :param inputs: An instance of :class:`DataPack` to be transformed.
+        :return: the output of all the transformed inputs.
+        """
+        if 'label' not in self._column_names:
+            raise ValueError(f"label is required from {inputs} \
+                             to generate pairs.")
+        aggregate_data = {}
+        for _, row in inputs.dataframe.iterrows():
+            # Note here the main id is set to be the id_left
+            curr_kid = row['id_left']
+            curr_label = row['label']
+            if curr_kid not in aggregate_data:
+                aggregate_data[curr_kid] = {}
+            if curr_label not in aggregate_data[curr_kid]:
+                aggregate_data[curr_kid][curr_label] = []
+            aggregate_data[curr_kid][curr_label].append(row)
+        all_pairs = []
+        for kid, label_info in aggregate_data.items():
+            label_list = sorted(label_info.keys(), reverse=True)
+            for hidx, high_label in enumerate(label_list[:-1]):
+                for high_sample in label_info[high_label]:
+                    low_samples = []
+                    for low_label in label_list[hidx+1:]:
+                        for low_sample in label_info[low_label]:
+                            low_samples.append(low_sample)
+                    curr_pairs_num = len(low_samples) // self._num_neg
+                    for idx in range(curr_pairs_num):
+                        beg = idx * self._num_neg
+                        end = (idx + 1) * self._num_neg
+                        all_pairs.append((high_sample, low_samples[beg:end]))
```

Comment:
we need to re-write this part, you used 7 loops (6 for loop + 1 sorting) from Line 80- Line 92, which is extremely inefficient.

Possible solution is
1. think of another data structure to store the data.
2. Use vectorized python to replace the for loop.

Reply:
Here, the function `transform_data` do the same thing with `make_pair`, but use an different function name.

Reply:
@bwanglzu The pair-wise generator need to make pairs according to the labels for each query. Firstly, we need to identify all documents associated with the same query, this is what `aggregate_data` do. Then, we need to sort the documents according to the corresponding label, in this way, all samples are organized in an descending order based on labels,  such as '2 1 1 0 0 0`. Finally, we make pairs for all the possible combinations, such as '[(2, 1), (2, 1), (2, 0), (2, 0), (2, 0), (1, 0), (1, 0), (1, 0), (1, 0), (1, 0), (1, 0)]'. That's why we need so many loops.

Reply:
@faneshion 
1: `df.groupby("query")`
2. `df.sort("label", ascending=False)`
3. Split the dataframe by labels (say label = 0 or 1) into 2 sub-dataframes.
4. for both child dataframes, call df.sample(n, replace=True)

Reply:
@bwanglzu So all the current data manipulations are based on pandas functions?
(1+num_neg)*k=batch_size,
df.sample(k, replace=True)
df.sample(num_neg*k, replace=True)
This part is a bit confuse. If a query has 1 positive candidate and num_neg negative candidates, we can make num_neg pairs. So should it be (num_pos*num_neg)*k=batch_size ?

Reply:
@yangliuy `num_pos`  alywas = 1, while `num_neg` is tunable, for example, dssm train with 1 pos together with 4 negative.